### ISSUE TITLE
Add support for a more robust detox config

### ISSUE THREAD

**Issue by simonracz** [wix/Detox#175]
The feature is described in a google docs [here](https://docs.google.com/document/d/1z2ENgzqNRHxxUfh81Illhf9Thjjgt3tGurPFg77bmtg/edit?usp=sharing)

**Comment by rumax** [comment]
Will it be possible to use system variables too?

For example
```
export DETOX_DEVICE="iPhone 5"
export DETOX_DEVICE_CI="iPhone 7"
```

**Comment by DanielMSchmidt** [comment]
Do we want to support a config file too? That could be used to configure the behavior more dynamically.

**Comment by rotemmiz** [comment]
the config can be loaded from anywhere, that's how we currently do it
```js
const config = require('../package.json').detox;
```

**Comment by DanielMSchmidt** [comment]
So ENV variables could be used there to overwrite, right?

**Comment by rotemmiz** [comment]
not sure what you mean, ENV vars from where ?

**Comment by DanielMSchmidt** [comment]
Sorry 🙈  @rumax mentions using environment variables to overwrite parts of the config. This overwrite part could be used in the generated file that includes `const config = require('../package.json').detox;`

By the way, what would you think of using the [jest-config]() package. That way we would use a way that might feel known and we could reuse that module. But that could be done separate from changing the structure of the config. 

I would now go ahead and try to build a first draft of the config described in the doc and we can think about ways to configure in the meantime 👍

**Comment by DanielMSchmidt** [comment]
After reading the proposal again I would like to add a `name` field on the topLevel, that way we could have better defaults for the binaryPaths. I would default the`name` with the name field from the package.json. That way we could have almost zero configuration (except simulators, but we could maybe default to the first one in the list there too, to be zero config by default)

**Comment by DanielMSchmidt** [comment]
@simonracz Regarding preparing the `parallel` feature, would you suggest to run multiple simulators in parallel that run the same test on each simulator (e.g. testing iPad & iPhone side by side) or would you rather split up the tests to run the suite faster (e.g. two iPhones 5 side by side) Maybe we could open up a new issue to keep track of the ideas in this direction.

**Comment by rotemmiz** [comment]
The first approach is pretty easy (testing the same on multiple devices), but that's not what I think will bring much value.
Being able to shard tests into multiple identical devices is a much more complex task (especially if we want to to keep the API as is), I think that its value is much bigger. We already have test suites that take over 10 minutes, I believe that parallelizing tests can slice that by more than half.

**Comment by simonracz** [comment]
Take this google docs as a high level proposal. Probably, there are things in it that even I don't agree with anymore.

I still think detox should have a more robust config.

Especially when,
1. Android support will be officially released
2. Detox will support multiple emulators/simulators/devices.

**Comment by DanielMSchmidt** [comment]
@simonracz Agreed! I would love to do some code around this, just to see if the envisioned core concept will work and to discuss then how to improve further from this. I think the inheritance / default values are a nice to add on top feature, that should come in another PR, I guess.

**Comment by simonracz** [comment]
Feel free to take over this. I am not working on this at the moment.

**Comment by LeoNatan** [comment]
@DanielMSchmidt Has this been resolved? I haven't followed it or the PR, but I see the PR is closed?

**Comment by DanielMSchmidt** [comment]
I think @rotemmiz wanted to give it a try, at least as far as I know

**Comment by LeoNatan** [comment]
@noomorph This, at least partially, will be supported once the full blown arguments PR lands, right?

**Comment by noomorph** [comment]
Partially — yes, for sure. @simonracz wrote quite an interesting doc, and we can take some ideas from it. Let's link to #1238.

**Comment by LeoNatan** [comment]
Closing this in favor of #1238
